File: usememory/map/applied/map_index/two_sum.java
// brute force
public int[] twoSum(int[] nums, int target){
  for (int i = 0; i < nums.length; i++){
    for (int j = i + 1; j < nums.length; j++){
      if (nums[j] == target - nums[i]){
        return new int[] { i , j };
      }
    }
  }
  throw new IllegalArgumentException("no two sum solution");
}

// two pass hash map
public int[] twoSum(int[] nums, int target){
  Map<Integer, Integer> numsMap = new HashMap<>();
  for (int i = 0; i < nums.length; i++){
    numsMap.put(nums[i], i);
  } 
  for (int i = 0; i < nums.length; i++){
    int complement = target - nums[i];
    // need to check not getting the same value twice i.e. if target
    // is 8 we get 4 and 8's complement 4 again
    if (numsMap.containsKey(complement) && numsMap.get(complement) != i){
      return new int[] { numsMap.get(complement), i};
    }
  }
  throw new IllegalArgumentException("no two sum solution");
}

// one pass hash map
public int[] twoSum(int[] nums, int target){
  Map<Integer, Integer> numsMap = new HashMap<>();
  for (int i = 0; i < nums.length; i++){
    int complement = target - nums[i];
    if (numsMap.containsKey(complement) && numsMap.get(complement) != i){
      return new int[] { numsMap.get(complement), i};
    }
    numsMap.put(nums[i], i);
  }
  throw new IllegalArgumentException("no two sum solution");
}

File: data_structures/map/applied/map_index/two_sum.scala
object Solution {
  // one pass hash map
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    val numsMap = scala.collection.mutable.HashMap[Int, Int]()

    for (i <- nums.indices) {
      val complement = target - nums(i)
      if (numsMap.contains(complement))
        return Array(numsMap(complement), i)
      else
        numsMap.put(nums(i), i)
    }
  }

  // one line 
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    nums.zipWithIndex
      .foldLeft(Map[Int, Int]()){
        case (numsMap, numsTuple @(x, complementIndex)) => 
          numsMap.get(target - x)
                 .map(index => return Array(index, complementIndex))
                 .getOrElse(numsMap + numsTuple)
      }
  }

  import scala.collection.immutable.HashMap
  def twoSum(nums: Array[Int], target: Int): Array[Int] = {
    @annotation.tailrec
    def recursiveTwoSum(map: HashMap[Int, Int], i: Int): Array[Int] = {
      val complement = target - nums(i)
      map.get(complement) match {
        case Some(complementIndex) => Array(complementIndex, i)
        case None => recursiveTwoSum(map + (nums(i) -> i), i + 1)
      }
    }
    recursiveTwoSum(HashMap(), 0)
  }
}